Enhance quota exceeded logging for admins

joshtrichards opened the pull request:
* Resolves: #37519 <!-- related github issue -->

## Summary

Enhances quota exceeded server-side logging for administrators by adding:
* target folder owner name (i.e. when shared)
 * indicator of user quota (vs disk space) issue

Context: The existing `Insufficient space` error logging doesn't indicate the folder owner (which is important since the sharee path is what's logged). In addition, quota issues are addressed differently than, say, disk space issues so administrators benefit from quickly noting the type of `Insufficient space` issue occurring.

Applies to file/folder operations: uploads, moves, copies

Also:
* improves - indirectly - the user-facing messaging when uploading (only*)
* Doesn't include the owner when it's useless (i.e. not shared)

Addresses backend portion of #37519 

*Currently working on independent (but related) PRs to unify frontend messaging with same target folder owner inclusion (when applicable)

## TODO

- [x] Review code

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- Screenshots before/after for front-end changes
- Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

joshtrichards committed:
Enhance quota exceeded logging for admins

* Adds owner of target folder (i.e. when shared)
* Only log owner if not current user
* Clarifies user quota vs disk space issue
* Special $path handling for non-Chunked-v2 uploads
* Mimic current path behavior for targets that already exist
* Add db access to unit tests

Addresses backend portion of #37519

Signed-off-by: Josh Richards <josh.t.richards@gmail.com>

joshtrichards commented:
For the record...

Existing log entry for all quota issues:

`Insufficient space in /Project/Videos/lightonflux, 27136230 required, 343725 available`

New (where owner is applicable): 

`Quota exceeded in /Project/Videos/lightonflux (owner: sarah), 27136230 required, 343725 available`

New (where owner is inapplicable but there's still a quota issue): 

`Quota exceeded in /QuotaTest, 341471232 required, 301045000 available`

joshtrichards commented on apps/dav/lib/Connector/Sabre/QuotaPlugin.php:
```diff
@@ -202,12 +202,17 @@ public function checkQuota($path, $length = null) {
 				}
-				throw new InsufficientStorage("Insufficient space in $path, $length required, $freeSpace available");
+				if ($targetOwner !== \OC_User::getUser()) {
+					throw new InsufficientStorage("Quota exceeded in $path (owner: $targetOwner), $length required, $freeSpace available");
```
Not seeing this, but I assume I'm overlooking something: can you clarify?

joshtrichards replied:
@blizzz `$targetOwner` is just a string that has the result of `$this->view->getOwner($path)` from just above that (and already contains the UID).

I'm still getting familiar with the code base so I may be missing something seemingly silly. Let me know. Thanks!

blizzz replied:
Somehow I though it returns an IUser, but you are right, sorry for this.

szaimen commented:
/rebase

szaimen commented:
tests failing...

joshtrichards commented:
Fixed (at least for me... we'll have to see what the tests say).

In the quota plugin, depending on which `before*` operation is being called, sometimes `checkQuota()` gets handed a `$path` that includes the basename of the URI and other times not so much.

That's current behavior so I'm leaving it be, but I added a check for non-chunked uploads in `checkQuota()` now just before the `getOwner()` call in this PR. That way we always gets a workable`$path` for non-chunked uploads (and, more importantly, the same as we'd have gotten if it had been a chunked upload)

The main reason non-chunked uploads triggered it is because the PUT operation never finishes (since the quota is exceeded) so there's really no file to check ownership of - all we have is the intended folder.

All my manual testing was apparently based on doing >10 MB uploads. And the way that chunking-v2 uploads work, the MOVE was giving us what we expected for `$path`. But <10 MB uploads just do a straight PUT and we receive the full path with the filename intact for the quota check. The integration tests failing all seem to be <10 MB (non-chunked).

joshtrichards commented:
Incidentally this also fixes a "cosmetic" bug as a side effect versus current master behavior where hitting quota with uploads in the root folder were being mishandled in the logging/notifications:

Before:
<img width="560" alt="NC Files - Quota Full Handling - Screenshot 2023-05-16 130845" src="https://github.com/nextcloud/server/assets/1731941/26d3146a-66b9-4209-a464-cd07c9f756ca">
After:
<img width="368" alt="NC Files - Quota Full Handling After - Screenshot 2023-05-16 131039" src="https://github.com/nextcloud/server/assets/1731941/4f5d1a01-cc6b-4521-9244-c6c3f139c641">

I'd noted this during various bits of testing but hadn't gotten around to checking for prior reports of or root causes for - so that's a nice benefit to this change even if it wasn't the target of (or caused by) this PR. Woohoo!

joshtrichards commented:
Tested WAY better with changes made earlier today, but needed an additional workaround for the lack of destination path existence handling in  `beforeCreateFile()`... 

(Now we'll see what the latest tests say *this* time.)

Workaround wasn't needed previously because only use for the path was for `getFreeSpace()`. I guess it isn't particularly choosy about whether the target exists or is a folder vs file... (presumably as long as the parent folder does exist it's okay since that's how it's been functioning to date in released versions...).

If this passes all tests, may want to consider a final refactor to shift the existence check up into `beforeCreateFile()` so things are consistent with `beforeCopy()` & `beforeMove()` (which seem to already do a reasonable job adjusting the passed along path to checkQuota() depending on the target destination's pre-existence or lack thereof).

But I'm hesitant to mess with it much within the confines of *this* PR, since it's a very important code pathway and ends up being way beyond the original scope I'd planned.

Gotta say this turned out to be a bit more work than I originally anticipated for a logging adjustment. It initially seemed quite easy. 😂

joshtrichards commented:
@szaimen When you get a chance can you approve the test workflow trigger? Looks like it's mad at me. :) Want to see if it passes.

joshtrichards commented:
Last full test run:

* Cypress: failing because it's running against my tree (I guess?)
* Performance-8.0: I suspect for similar reasons given some interaction with the 3rdparty modules handling (maybe #35192 still not fully fixed or perhaps because I happened to upload from a working space where I hadn't done an init on 3rdparty submodules [latter fixed just now])
  - https://github.com/nextcloud/server/actions/runs/4997640538/jobs/9051783004?pr=37581
* Drone
  - nodb: "Your test case is not allowed to access the database" I mean... yeah, right? 🤔 
    - https://drone.nextcloud.com/nextcloud/server/34205/9/4
  - integration-contacts-menu: Looks to bge #38019 & #38337

blizzz commented:
> nodb: "Your test case is not allowed to access the database" I mean... yeah, right? thinking 

changed code accesses database now, `apps/dav/tests/unit/Connector/Sabre/QuotaPluginTest.php` will have to be equipped with "@group DB" in the class annotation.

skjnldsv commented:
What's the status here @joshtrichards :)
